File: DCP-1836/solution.swift

// Problem 1836 (Code examples of Polymorphism)

import Foundation

// ad-hoc polymorphism
func sum(_ a: Int, _ b: Int) -> Int {
    return a + b
}

func sum(_ a: String, _ b: String) -> String {
    return a + b
}

// Usage
let numberResult = sum(3, 5)      // Returns 8
let stringResult = sum("ad", "hoc") // Returns "adhoc"

// parametric polymorphism
struct CargoHold<T> {
    var items: [T] = []
    
    mutating func addItem(_ item: T) {
        items.append(item)
    }
}

// Usage
var intCargoHold = CargoHold<Int>()
intCargoHold.addItem(100)

var stringCargoHold = CargoHold<String>()
stringCargoHold.addItem("Medical Supplies")

class Shape {
    func draw() {
        print("Drawing a shape")
    }
}

class Circle: Shape {
    override func draw() {
        print("Drawing a circle")
    }
}

// Usage
func drawShape(shape: Shape) {
    shape.draw()
}

drawShape(shape: Circle())  // Prints "Drawing a circle"

File: DCP-1836/solution.kt
// Problem 1836

// ad-hoc polymorphism
fun doubleof(a: Int): Int {
    return 2 * a
}

fun doubleof(a: Double): Double {
    return 2.00 * a
}

// Usage
println(doubleof(4))    // Prints 8
println(doubleof(4.3))  // Prints 8.6

// parametric polymorphism
class Box<T> {
    private var content: T? = null
    
    fun put(item: T) {
        content = item
    }
    
    fun get(): T? {
        return content
    }
}

// Usage
val numberBox = Box<Int>()
numberBox.put(42)
val stringBox = Box<String>()
stringBox.put("Hello Generics")

// subtyping polymorphism
open class Sup {
    open fun method1() {
        println("printing method 1 from inside Sup")
    }
}

class Sum: Sup() {
    override fun method1() {
        println("printing method 1 from inside Sum")
    }
}

// Usage
val b = Sum()
b.method1()  // Prints "printing method 1 from inside Sum"